### PR TITLE
Financial ACLs shouldn't restrict financial accounts by relationship type

### PR DESCRIPTION
**Pull request by MegaphoneJon**
Overview
----------------------------------------
Financial ACLs restrict viewing financial accounts to those with a relationship of "Income Account is".  This is true even for super-admins.

So if you want to make an API call to get all, say, Expense accounts, you can't.

To replicate:
* Run this API call on a demo site (as super-admin) with Financial ACLs disabled:
```php
$financialAccounts = \Civi\Api4\FinancialAccount::get(FALSE)->execute();
```
Enable Financial ACLs and run again, still as super-admin.  

Before
----------------------------------------
Get 4 results.

After
----------------------------------------
Get 9 results.

Comments
----------------------------------------
Advanced mathematicians might notice that 9 != 14.  There are some accounts that, by default, aren't tied to any financial type. IMO those should also be returned, but I'm not sure if that's as clear-cut.  Would appreciate feedback from those with deep knowledge (e.g. @JoeMurray) to see if there's any reason why, e.g., "Premiums Inventory" or "Discounts" account types shouldn't appear in an API call when Financial ACLs are enabled.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
Civi\Financialacls\FinancialAccountTest::testGetFinancialAccount
Failed asserting that actual size 4 matches expected size 1.

/home/homer/buildkit/build/build-2/web/sites/all/modules/civicrm/ext/financialacls/tests/phpunit/Civi/Financialacls/FinancialAccountTest.php:25
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2307

**Comment by JoeMurray** [comment]
TL;DR this is not necessary and can lead to data leakage. 

I agree there is at least one problem here regarding how CheckPermissions and super-admin user overrides of ACLs/permissioning should work (see B and C below).

A) Correction to overview: The current logic when Financial ALCs are enabled is that all financial accounts that are in any EntityFinancialAccount relationship with a viewable financial type are viewable (not just 'Income Account Is'). Note that there are more of these types of relationships than appear in the default tarball data:


The justification for this is that enabling Financial ACLs means that there is a desire to hide certain financial information from some users. The paradigm is to only show what is explicitly enabled for viewing. If an org configures separate A/R and expense and banking expense financial accounts for each Financial Type, this allows all transactions for each Financial Type to be hidden properly without leakage. 

Potential leakage in terms of people viewing information that they shouldn't can occur in many ways. 

Avoiding leakage requires that all of the financial accounts that are in an EntityFinancialAccount relationship with an unviewable financial type should not be viewable. Bad configuration in the form of shared financial accounts between viewable and unviewable financial types can lead to leakage.

For example, sharing a single Accounts Receivable Account across all financial types when creating A/R entries for all contributions would allow the A/R entries to show contribution information to people who are not permissioned to view the revenue account/contribution data. One way to avoid this is a Chart of Accounts with subaccounts for the A/R account for each of the revenue accounts, with separate CiviCRM A/R financial accounts for each. A convenient workaround if these A/R subaccounts don't exist in the external accounting system is just to configure separate A/R Financial Accounts in CiviCRM for each Financial Type, and map them all to the same A/R in the accounting system.

Similarly, leakage could occur based on expense or cost of sales accounts, and can be avoided through configuration.

Multiplying the A/R or expense accounts in this way leads to inconvenience and some confusion when using the core reports. But it should be possible to coalesce the data appropriately using SearchKit.

I don't have a strong view about the viewability of accounts that are not in an EntityFinancialAccount relationship with any financial type. I tend to think it is safer to not have them viewable, but I won't insist.

But letting financial accounts be visible so long as they are not Income Account Is accounts for unviewable financial types can lead to leakage. So I'm not in favour of this PR being merged.

B) When Check Permissions is false, shouldn't all data be returned? Isn't that how CheckPermissions is interpreted in other API contexts? So \Civi\Api4\FinancialAccount::get(FALSE)->execute(); should return all Financial Accounts no matter which user is calling and no matter what permissions have been set when Financial ACLs extension is enabled. This is not subject of current PR, and so it's probably best to let sleeping dogs lie.

C) When user is super-admin, shouldn't all data be returned even if Check Permissions is true? Isn't that how super-admin is interpreted in other API contexts? So when executed by super-admin user, \Civi\Api4\FinancialAccount::get(TRUE)->execute(); should return all Financial Accounts. This is not subject of current PR, and so it's probably best to let sleeping dogs lie.

----
Does this make sense, @MegaphoneJon and @eileenmcnaughton ?

**Comment by MegaphoneJon** [comment]
@JoeMurray Do you have thoughts on my last message?  I can just do the other PR you suggest but I want to make sure I'm thinking about the security correctly.

**Comment by JoeMurray** [comment]
Sorry for the delay. Here's response on the easy parts B and C: yes please go ahead.

I will come back later to respond to your questions about A when I have a few hours to test and answer.

**Comment by MegaphoneJon** [comment]
@JoeMurray Adding/editing payment processors is broken on Civi 6.1+, so I implemented part "C" as you requested.  Part B already works as described - maybe that got fixed in the past year?

Test fail looks unrelated.

File: ext/financialacls/financialacls.php [hunk 1]
```diff
@@ -85,6 +85,9 @@ function financialacls_civicrm_pre($op, $objectName, $id, &$params) {
  * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_selectWhereClause
  */
 function financialacls_civicrm_selectWhereClause($entity, &$clauses) {
+  if (CRM_Core_Permission::check([['administer CiviCRM', 'administer CiviCRM Financial Types']])) {
```
Comment (eileenmcnaughton) [inline ext/financialacls/financialacls.php:88]: Super admin is actually 'all CiviCRM permissions and ACLs'

image

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon We just looked at this & it should be mergeable if you change the 'super admin' to the right permission - but does it fully fix the regression?

**Comment by MegaphoneJon** [comment]
@eileenmcnaughton I implemented the requested change.  I assume "Administer CiviCRM" has "Administer CiviCRM Financial Types" as an implied permission, so that should be fine.

This doesn't fully fix the regression - it still affects anyone with the "Edit Payment Processors" permission who doesn't have one of the two permissions listed in this PR.  But expanding it to include them with let data leak.  Joe asked Monish to look into the solution I proposed.

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon 'Administer CiviCRM' is a lower permission that the Super permission - it deliberately does NOT override ACLs - so people do need to add a permission to override ACLs

**Comment by eileenmcnaughton** [comment]
@JoeMurray so I think we are waiting for more feedback from you / @monishdeb on how to deal with the situation where people with the permission to "Edit Payment Processors" do not in fact have the ability to do that....
